src/windows/views/timeline_webview.py

```diff
@@ -1998,6 +1998,15 @@ def Volume_Triggered(self, action, clip_ids, position="Entire Clip"):
         """Callback for volume context menus"""
         log.info(action)
 
+        # Callback function, to redraw audio data after an update
+        def callback(self, clip_id, callback_data):
+            has_audio_data = callback_data
+            print('has_audio_data: %s' % has_audio_data)
+
+            if has_audio_data:
+                # Re-generate waveform since volume curve has changed
+                self.Show_Waveform_Triggered(clip_id)
+
```

Comment:
Why not put this right before the `run_js()` call it's passed to? There's no reason it needs to be defined right at the top of the function; seems like it just makes the code harder to read.

Ooh, hey — ­if `self` is only used to look up the function the _callback_ is going to call, could you just pass that directly, instead? Then the callback can be just a simple function, instead of looking like a method. I'm thinking something like:
```python3
# ...prior stuff in function...
def cb(clip_id, fn, result):
    if result:
        fn(clip_id)
self.run_js(JS_SCOPE_SELECTOR + ".hasAudioData('{}');".format(clip.id),
    partial(cb, clip.id, self.Show_Waveform_Triggered))
```
I have to test that, but it _should_ work.

Reply:
Hrm, wait, that may not work — now I'm thinking that it loses the `self` argument to `Show_Waveform_Triggered`.

Reply:
Oh, hey, it does work! Python is smart enough to know that `self.function_name`, when passed as a function object, effectively represents a call to `function_name(self, ...)` — the value of `self` doesn't need to be explicitly carried alongside it:
```python3
Python 3.8.3 (default, May 29 2020, 00:00:00) 
[GCC 10.1.1 20200507 (Red Hat 10.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import functools
>>> 
>>> class Thinger:
...     def thing(self, count=0):
...         print(["thing! " for x in range(0,count)])
...     def indirect(self, do_thingy=False):
...         def cb(fn, do_it, val):
...             if (do_it):
...                 fn(val)
...         return functools.partial(cb, self.thing, do_thingy)
... 
>>> a = Thinger()
>>> doit = a.indirect(True)
>>> doit(2)
['thing! ', 'thing! ']
>>> doit(5)
['thing! ', 'thing! ', 'thing! ', 'thing! ', 'thing! ']
```